In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import cross_validate
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn import datasets
from sklearn.metrics import accuracy_score
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.linear_model import LogisticRegression

In [3]:
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/data/classify_imb.csv')

In [4]:
data

A         B         C         D         E         F         G  \
0    2.868690 -0.106937  1.065444  0.749917  2.138981 -0.267874 -4.287288   
1   -1.463642 -1.291157 -1.464761  1.287344 -1.994637  0.831797  0.130555   
2   -2.208983  1.112905  1.484842  1.455104 -0.380712 -0.465900  3.464477   
3    1.757493 -0.194035 -1.502209 -1.227620 -1.629453  2.192913 -0.774430   
4   -0.621503 -1.709654 -2.155096 -1.982270 -1.205284 -0.285097 -1.279000   
..        ...       ...       ...       ...       ...       ...       ...   
995 -0.597016 -0.058968 -0.883755 -1.185085  2.798069 -0.688958  1.403163   
996 -3.101287  1.255737 -0.995375  1.528169  1.289261  2.116091  1.573981   
997 -1.430878  1.423774 -2.381375 -1.064439  5.550441  0.302901  1.445181   
998  0.965698 -0.147079 -2.212334  1.209485 -2.267677  1.774173  0.446710   
999  1.315056 -1.547771  1.485064 -2.915247  0.720597 -1.693207 -2.003952   

            H         I         J  ...         L         M         N  \
0   -0.003229  0.102618 -1.594098  ... -0.031744 -1.836612  2.508588   
1   -0.555585 -1.746450  0.320191  ... -0.948864  0.497305  2.651037   
2    1.561576 -0.972615  2.646365  ...  0.571619 -0.573899 -4.393691   
3   -1.284792  0.433154  2.233775  ...  0.089484 -2.290146 -2.274138   
4   -3.831940 -1.205034 -4.213100  ...  0.808936  0.647993  4.629954   
..        ...       ...       ...  ...       ...       ...       ...   
995 -1.843103  0.346712 -2.234694  ... -0.785120  1.470929 -3.184477   
996  0.422076 -1.045971  1.904110  ... -0.370083 -0.902425 -1.788914   
997 -5.063313 -0.333850 -5.566545  ... -1.533786 -0.720187 -4.316099   
998 -2.363656 -0.795120  1.195916  ...  1.107171  2.354562 -2.371758   
999  2.610604 -0.218337 -1.528609  ... -0.350076  0.609276  5.141815   

            O         P         Q         R         S         T  Target  
0    2.123112 -1.599241  0.944821  0.330276  1.017247 -0.440207       0  
1    1.722998  0.290210 -1.278545  0.488835  1.577497 -1.097335       1  
2   -3.936344  5.499447  0.874778 -0.858103 -1.040617 -2.759917       0  
3   -3.454273 -0.904307  0.459121 -0.043155 -2.129675 -0.559314       0  
4    3.860843 -1.288296  0.992935  1.069965  3.676966  1.005724       0  
..        ...       ...       ...       ...       ...       ...     ...  
995 -3.113836 -0.166130  1.515417 -0.507200 -1.385448  1.099414       1  
996 -2.686736  1.720358 -0.780103 -1.834606  0.190024 -0.436017       0  
997 -3.982227 -2.949054  3.422676 -0.652477 -0.303372 -0.381482       1  
998 -3.317124 -1.024851  0.684339 -0.357698 -1.878043  1.143353       0  
999  6.492223 -4.050462 -2.719196  0.729706  1.067730 -0.331290       0  

[1000 rows x 21 columns]

In [5]:
data['Target'].sum()/len(data)

0.321

In [6]:
X=data.drop('Target',axis=1)
y=data['Target']

When you split your data into training and test sets, set random_state = 5. (2 points)

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3, random_state=5)

In [8]:
print(sum(y_train)/len(y_train),":",sum(y_test)/len(y_test))

0.3157142857142857 : 0.3333333333333333


Try a random forest (use random_state = 5 and 400 trees), a boosting classifier (use n_estimators = 300, learning rate =0.1, and max_depth =3), and a logistic regression classifier. (9 points: 3 for each classifier)




In [9]:
logreg=LogisticRegression().fit(X_train,y_train)

In [10]:
logreg.score(X_test,y_test)

0.6833333333333333

In [11]:
pd.crosstab(y_test,logreg.predict(X_test))

col_0     0   1
Target         
0       194   6
1        89  11

In [12]:
rf = RandomForestClassifier(n_estimators=400,max_features='sqrt',random_state=5).fit(X_train,y_train)

In [13]:
accuracy_score(rf.predict(X_test),y_test)

0.7766666666666666

In [14]:
pd.crosstab(y_test,rf.predict(X_test),rownames=['True'], colnames=['Predicted'])

Predicted    0   1
True              
0          190  10
1           57  43

In [15]:
digit_boost = GradientBoostingClassifier(n_estimators=300, learning_rate=0.1,max_depth=3, random_state=5).fit(X_train, y_train)

In [16]:
accuracy_score(digit_boost.predict(X_test),y_test)


0.7566666666666667

In [17]:
pd.crosstab(y_test,digit_boost.predict(X_test))

col_0     0   1
Target         
0       181  19
1        54  46

Which is best? How can you tell? (6 points)

Here the random forest is better than other models. Random Forest trains and treats each tree independently it also considers randomisation ie it uses a randomly sampled data. This further helps to make the model better and robust compared to the other models and does not overfit the traning data. 


Comment on the precision, recall, and accuracy of your model. (3 points)

In [18]:
from sklearn.metrics import classification_report

In [19]:
print(classification_report(y_test,digit_boost.predict(X_test)))

              precision    recall  f1-score   support

           0       0.77      0.91      0.83       200
           1       0.71      0.46      0.56       100

    accuracy                           0.76       300
   macro avg       0.74      0.68      0.69       300
weighted avg       0.75      0.76      0.74       300



In [20]:
print(classification_report(y_test,logreg.predict(X_test)))

              precision    recall  f1-score   support

           0       0.69      0.97      0.80       200
           1       0.65      0.11      0.19       100

    accuracy                           0.68       300
   macro avg       0.67      0.54      0.50       300
weighted avg       0.67      0.68      0.60       300



In [21]:
print(classification_report(y_test,rf.predict(X_test)))

              precision    recall  f1-score   support

           0       0.77      0.95      0.85       200
           1       0.81      0.43      0.56       100

    accuracy                           0.78       300
   macro avg       0.79      0.69      0.71       300
weighted avg       0.78      0.78      0.75       300



The Random forest model has the highest accuracy of 0.78, higest precision 0.77 and highest recall 0.95.
The recall for all the three model is high and the precision is low. High recall, low precision. Our classifier casts a very wide net, catches not only a lot of true positves, but also a lot of other things. Our classifier considers a lot of things as true positive but a lot of them were not true positive.